R/xgx_stat_ci.R

```diff
@@ -99,87 +99,150 @@
 #' @importFrom binom binom.exact
 #' @importFrom ggplot2 stat_summary
 #' @importFrom ggplot2 aes
-#' @importFrom ggplot2 position_dodge
+#' @importFrom ggplot2 position_dodeg
 #' @export
-xgx_stat_ci <- function(mapping = NULL, data = NULL, conf_level = 0.95,
-                        distribution = "normal",
-                        geom = list("point", "line", "errorbar"),
-                        position = "identity",
-                        fun.args = list(),
-                        na.rm = FALSE,
-                        show.legend = NA,
-                        inherit.aes = TRUE,
-                        ...) {
-  if (!(conf_level > 0.5 && conf_level < 1)) {
-    stop("conf_level should be greater than 0.5 and less than 1")
-  }
-
-  percentile_value <- conf_level + (1 - conf_level) / 2
-
-  conf_int <- function(y, conf_level, distribution) {
-    y <- stats::na.omit(y)
-
-    if (distribution == "normal") {
-      conf_int_out <- data.frame(
-        y = mean(y),
-        ymin = mean(y) - stats::qt(percentile_value,
-                                   length(y)) * sqrt(stats::var(y) / length(y)),
-        ymax = mean(y) + stats::qt(percentile_value,
-                                   length(y)) * sqrt(stats::var(y) / length(y))
+xgx_stat_ci <- function(
+                    data = NULL,
+                    mapping = NULL,
+                    geom = list("point", "line", "errorbar"),
+                    position = "identity",
+                    na.rm = FALSE,
+                    show.legend = NA, 
+                    inherit.aes = TRUE,
+                    
+                    conf_level = 0.95,
+                    distribution = "normal",
+                    bins = NULL,
+                    ...) {
+  lays <- list()
+  
+  for (igeom in geom) {
+    if (is.null(bins)) {
+      lay = layer(
+        stat = StatSummary,
+        data = data,
+        mapping = mapping,
+        geom = igeom,
+        position = position,
+        show.legend = show.legend,
+        inherit.aes = inherit.aes,
+        params = list(
+          fun.args = list(),
+          fun.data = function(y) xgx_conf_int(y, conf_level, distribution),
+          na.rm = na.rm,
+          ...)
       )
-    } else if (distribution == "lognormal") {
-      yy <- log(y)
-      conf_int_out <- data.frame(
-        y = exp(mean(yy)),
-        ymin = exp(mean(yy) - stats::qt(percentile_value, length(yy)) * sqrt(stats::var(yy) / length(yy))),
-        ymax = exp(mean(yy) + stats::qt(percentile_value, length(yy)) * sqrt(stats::var(yy) / length(yy)))
+    }
+    else {
+      lay = layer(
+        stat = StatCI,
+        data = data,
+        mapping = mapping,
+        geom = igeom,
+        position = position,
+        show.legend = show.legend,
+        inherit.aes = inherit.aes,
+        params = list(
+          conf_level = conf_level,
+          distribution = distribution,
```

Comment:
maybe can leave 

        params = list(
          fun.args = list(),
          fun.data = function(y) xgx_conf_int(y, conf_level, distribution),
          na.rm = na.rm,

so that it's customizable. Then the only thing that StatBin would do is to bin the x values and apply whatever function is in fun.data

Reply:
I agree with this and will work on implementing it.  
It is likely that the best solution to do this is to separate the ordinal functionality into a separate layer with a different ggproto (derived from the state of StatCI right now), which will only be called if the distribution is ordinal/multinomial/binomial.

Reply:
let's get the xgx_stat_ci working with the quantile bins first, then worry about ordinal after (since the ordinal part is still not quite there). Can you move all the ordinal stuff to Rdev, until it's working as we want it to?